contracts/governance/token/WrappedGovernanceToken.sol

```diff
@@ -0,0 +1,555 @@
+     */
+    function symbol() external view returns (string memory) {
+        return _symbol;
+    }
+
+    /**
+     * @dev The number of decimals is set at construction time and is equal to
+     * the underlying token's decimals, provided that the underlying token does
+     * not change that value.
+     * @return Decimals used by the wrapped governance token.
+     */
+    function decimals() external view override returns (uint8) {
+        return _decimals;
+    }
+
+    /**
+     * @param account Address to check the balance of.
+     * @return Balance of given account.
+     */
+    function balanceOf(address account) public view returns (uint256) {
+        return _balances[account];
+    }
+
+    /**
+     * @return Total supply of the wrapped governance token.
+     */
+    function totalSupply() public view returns (uint256) {
+        return _totalSupply;
+    }
+
+    /**
+     * @dev Creates `amount` tokens and assigns them to `account`, increasing
+     * the total supply.
+     */
+    function _mint(address account, uint256 amount) internal virtual {
+        require(account != address(0), "ERC20: mint to the zero address");
+
+        _totalSupply += amount;
+        _balances[account] += amount;
+
+        _moveDelegates(address(0), account, amount);
+
+        emit Transfer(address(0), account, amount);
+    }
+
+    /**
+     * @dev Destroys `amount` tokens from `account`, reducing the total supply.
+     */
+    function _burn(address account, uint256 amount) internal virtual {
+        require(account != address(0), "ERC20: burn from the zero address");
+
+        uint256 accountBalance = _balances[account];
+        require(accountBalance >= amount, "ERC20: burn amount exceeds balance");
+        _balances[account] = accountBalance - amount;
+        _totalSupply -= amount;
+
+        _moveDelegates(account, address(0), amount);
+
+        emit Transfer(account, address(0), amount);
+    }
+
+    /**
+     * @notice Deposit governance tokens.
+     * @param amount Number of governance tokens to deposit.
+     */
+    function deposit(uint256 amount) external override {
+        underlying.safeTransferFrom(msg.sender, address(this), amount);
+
+        UserProxy proxy = proxies[msg.sender];
+
+        // Deploy new user proxy if user does not have one.
+        if (address(proxy) == address(0)) {
+            proxy = new UserProxy();
+            proxies[msg.sender] = proxy;
+
+            emit UserProxyCreated(msg.sender, address(proxy));
+        }
+
+        // Mint same number of WGT tokens for user
+        _mint(msg.sender, amount);
+    }
+
+    /**
+     * @notice Withdraw governance tokens.
+     * @param amount Number of governance tokens to withdraw
+     */
+    function withdraw(uint256 amount) external override {
+        require(
+            availableBalanceOf(msg.sender) >= amount,
+            "WGT::withdraw: Insufficient balance"
+        );
+
+        underlying.safeTransfer(msg.sender, amount);
```

Comment:
function is not `nonReentrant`, don't you rather want to burn before transfer?

Reply:
Good point! 

For normal token cases this order doesn't hurt, but someone could wrap a malicious token that calls this contracts' methods during the transfer call, which *would* result in a reentrancy accounting issue.

Will change.